### ISSUE TITLE
Incorrect use of temporary paths, and dependency is missing a require

### ISSUE THREAD

**Issue by YorickPeterse** [zeisler/visualize_ruby#6]
This code:

```ruby
# frozen_string_literal: true

require 'visualize_ruby'

code = <<~RUBY
  class TodosFinder
    prepend FinderWithCrossProjectAccess
    include FinderMethods
    include Gitlab::Utils::StrongMemoize

    requires_cross_project_access unless: -> { project? }

    NONE = '0'.freeze

    ISSUABLE_TYPES = Set.new(%w(Issue MergeRequest Epic)).freeze

    attr_accessor :current_user, :params

    def initialize(current_user, params = {})
      @current_user = current_user
      @params = params
    end

    def execute
      items = current_user.todos
      items = by_action_id(items)
      items = by_action(items)
      items = by_author(items)
      items = by_state(items)
      items = by_type(items)
      items = by_group(items)
      # Filtering by project HAS TO be the last because we use
      # the project IDs yielded by the todos query thus far
      items = by_project(items)

      sort(items)
    end

    private

    def action_id?
      action_id.present? && Todo::ACTION_NAMES.key?(action_id.to_i)
    end

    def action_id
      params[:action_id]
    end

    def to_action_id
      Todo::ACTION_NAMES.key(action.to_sym)
    end

    def action?
      action.present? && to_action_id
    end

    def action
      params[:action]
    end

    def author?
      params[:author_id].present?
    end

    def author
      strong_memoize(:author) do
        if author? && params[:author_id] != NONE
          User.find(params[:author_id])
        end
      end
    end

    def project?
      params[:project_id].present?
    end

    def group?
      params[:group_id].present?
    end

    def project
      strong_memoize(:project) do
        if project?
          project = Project.find(params[:project_id])

          project unless project.pending_delete?
        end
      end
    end

    def group
      strong_memoize(:group) do
        Group.find(params[:group_id])
      end
    end

    def type?
      type.present? && ISSUABLE_TYPES.include?(type)
    end

    def type
      params[:type]
    end

    def sort(items)
      params[:sort] ? items.sort_by_attribute(params[:sort]) : items.order_id_desc
    end

    def by_action(items)
      action? ? items.where(action: to_action_id) : items
    end

    def by_action_id(items)
      action_id? ? items.where(action: action_id) : items
    end

    def by_author(items)
      author? ? items.where(author_id: author.try(:id)) : items
    end

    def by_project(items)
      project? ? items.where(project: project) : items
    end

    def by_group(items)
      return items unless group?

      groups = group.self_and_descendants
      project_todos = items.where(project_id: Project.where(group: groups).select(:id))
      group_todos = items.where(group_id: groups.select(:id))

      union = Gitlab::SQL::Union.new([project_todos, group_todos])

      Todo.from("(\#{union.to_sql}) \#{Todo.table_name}")
    end

    def by_state(items)
      params[:state].to_s == 'done' ? items.done : items.pending
    end

    def by_type(items)
      type? ? items.where(target_type: type) : items
    end
  end
RUBY

VisualizeRuby.new do |vb|
  vb.ruby_code = code

  vb.trace('TodosFinder.new.execute')

  vb.output_path = 'trace.png'
end
```

Produces this error:

```
/home/yorickpeterse/.gem/ruby/2.4.4/gems/dissociated_introspection-0.9.1/lib/dissociated_introspection/ruby_class.rb:4:in `<class:RubyClass>': uninitialized constant DissociatedIntrospection::RubyClass::Forwardable (NameError)
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/dissociated_introspection-0.9.1/lib/dissociated_introspection/ruby_class.rb:2:in `<module:DissociatedIntrospection>'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/dissociated_introspection-0.9.1/lib/dissociated_introspection/ruby_class.rb:1:in `<top (required)>'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/dissociated_introspection-0.9.1/lib/dissociated_introspection.rb:8:in `<top (required)>'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/builder.rb:1:in `<top (required)>'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby.rb:8:in `<top (required)>'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:135:in `require'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:135:in `rescue in require'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	from /tmp/viz.rb:3:in `<main>'
```

If we add `require 'forwardable'` at the start, we instead get this error:

```
/home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/2.4.0/tempfile.rb:133:in `initialize': No such file or directory @ rb_sysopen - /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/tmp/ruby_file20180906-23667-1lz3364.rb (Errno::ENOENT)
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/2.4.0/tempfile.rb:133:in `open'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/2.4.0/tempfile.rb:133:in `block in initialize'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/2.4.0/tmpdir.rb:132:in `create'
	from /home/yorickpeterse/.rubies/ruby-2.4.4/lib/ruby/2.4.0/tempfile.rb:131:in `initialize'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/execution_tracer.rb:45:in `new'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/execution_tracer.rb:45:in `temp_file'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/execution_tracer.rb:61:in `file!'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/execution_tracer.rb:53:in `ruby_file'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/execution_tracer.rb:22:in `trace'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/runner.rb:32:in `highlight_trace'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby/runner.rb:20:in `run!'
	from /home/yorickpeterse/.gem/ruby/2.4.4/gems/visualize_ruby-0.11.0/lib/visualize_ruby.rb:20:in `new'
	from /tmp/viz.rb:159:in `<main>'
```

This seems to be caused by this line (in `execution_tracer.rb`):

```ruby
file        = Tempfile.new(%w(ruby_file .rb), File.expand_path(File.join(File.dirname(__FILE__), "../../tmp")))
```

This should probably just use something like `Tempfile.new(..., Dir.tmpdir)`

**Comment by zeisler** [comment]
This has been dealt with in https://github.com/zeisler/visualize_ruby/commit/bcae0aea45e4aa1ab32372e0f035e9a9f08204a4#diff-755adc809a7abb3a89cabe761e43f05f and released in version 0.15.1

**Comment by YorickPeterse** [comment]
@zeisler Thanks!

**Comment by x20mar** [comment]
Hey @zeisler 

I'm still getting this issue. You mentioned that this is fixed in version 0.15.1 however, [rubygems](https://rubygems.org/gems/visualize_ruby) does not have this version.

Its last release is 0.11, published in July 2018.

Suggestions?

**Comment by zeisler** [comment]
@x20mar v0.16.0 has just been released onto Rubygems, enjoy!
